pkg/ui/terminal.go

```diff
@@ -43,34 +51,124 @@ func NewTerminalUI(journal journal.Recorder) (*TerminalUI, error) {
+
+	return u, nil
 }
 
-func (u *TerminalUI) RenderOutput(ctx context.Context, s string, styleOptions ...StyleOption) {
-	log := klog.FromContext(ctx)
+func (u *TerminalUI) Close() error {
+	var errs []error
+	if u.subscription != nil {
+		if err := u.subscription.Close(); err != nil {
+			errs = append(errs, err)
+		} else {
+			u.subscription = nil
+		}
+	}
+	return errors.Join(errs...)
+}
 
-	u.journal.Write(ctx, &journal.Event{
-		Action: journal.ActionUIRender,
-		Payload: map[string]any{
-			"text": s,
-		},
-	})
+func (u *TerminalUI) DocumentChanged(doc *Document, block Block) {
+	blockIndex := doc.IndexOf(block)
+
+	if blockIndex != doc.NumBlocks()-1 {
+		klog.Warningf("update to blocks other than the last block is not supported in terminal mode")
+		return
+	}
+
+	if u.currentBlock != block {
+		u.currentBlock = block
+		if u.currentBlockText != "" {
+			fmt.Printf("\n")
+		}
+		u.currentBlockText = ""
+	}
+
+	text := ""
+
+	var styleOptions []StyleOption
+	switch block := block.(type) {
+	case *ErrorBlock:
+		styleOptions = append(styleOptions, Foreground(ColorRed))
+		text = block.Text()
+	case *FunctionCallRequestBlock:
+		styleOptions = append(styleOptions, Foreground(ColorGreen))
+		text = block.Text()
+	case *AgentTextBlock:
+		styleOptions = append(styleOptions, RenderMarkdown())
+		if block.Color != "" {
+			styleOptions = append(styleOptions, Foreground(block.Color))
+		}
+		text = block.Text()
+	case *InputTextBlock:
+		fmt.Print("\n>>> ")
+		reader := bufio.NewReader(os.Stdin)
+		query, err := reader.ReadString('\n')
+		if err != nil {
+			block.Observable().Set("", err)
+		} else {
+			block.Observable().Set(query, nil)
+		}
+		return
+
+	case *InputOptionBlock:
+		fmt.Printf("%s\n", block.Prompt)
+
+		for {
+			fmt.Print("  Enter your choice (number): ")
+			var response string
+			_, err := fmt.Scanln(&response)
+			if err != nil {
+				block.Observable().Set("", err)
+				break
+			}
+
+			choice := strings.TrimSpace(response)
+
+			if slices.Contains(block.Options, choice) {
+				block.Observable().Set(choice, nil)
+				break
+			}
+
+			// If not returned, the choice was invalid
+			fmt.Printf("  Invalid choice. Please enter one of: %s\n", strings.Join(block.Options, ", "))
+			continue
+		}
+		return
+	}
 
-	computedStyle := &style{}
+	computedStyle := &ComputedStyle{}
 	for _, opt := range styleOptions {
 		opt(computedStyle)
 	}
 
-	if computedStyle.renderMarkdown {
-		out, err := u.markdownRenderer.Render(s)
-		if err != nil {
-			log.Error(err, "Error rendering markdown")
+	printText := text
+
+	if u.currentBlockText != "" {
+		if strings.HasPrefix(text, u.currentBlockText) {
+			printText = strings.TrimPrefix(printText, u.currentBlockText)
+		} else {
+			klog.Warningf("text did not match text already rendered; text %q; currentBlockText %q", text, u.currentBlockText)
 		}
-		s = out
 	}
+	u.currentBlockText = text
+
+	// TODO: Reintroduce markdown support (it's difficult with streaming)
```

Comment:
Does this mean, if we merge this, the current terminal UI won't be rendering markdown ?